src/mlmc/quantity.py

```diff
@@ -788,16 +881,35 @@ def size(self) -> int:
     def base_qtype(self):
         return self._qtype.base_qtype()
 
-    def replace_scalar(self, new_qtype):
+    @staticmethod
+    def replace_scalar(original_qtype, substitute_qtype):
```

Comment:
Convert to proper method of QType.
This can be much simpler if implemented by recursive construction of the replaced type. Returning the new instance. Use separate implementation for dict and scalar instead of
`isinstance`.

Reply:
I think convert to proper method of QType is not good idea. I agree with separate implementation for dict and scalar.

I created new issue.